Include TensorExpression interface in Tensor.hpp and clean up TensorExpression includes

macedo22 opened the pull request:
## Proposed changes

This PR adds to `Tensor.hpp` the `include`s necessary for interfacing with `TensorExpression`s - everything that's needed to write a tensor equation with `TensorExpression`s. As evidence that the `include`s appear to work properly, this PR also updates the `TensorExpression` tests that write tensor equations to simply include `Tensor.hpp` in lieu of including individual `DataStructures/Tensor/Expressions/....hpp` files for the generic indices, operations, and `evaluate`.

The motivation for this change is to have `#include DataStructures/Tensor/Tensor.hpp` provide all the ingredients necessary for writing a tensor equation with `TensorExpression`s:
- the generic indices (`TensorIndex`s)
- the operations (`+`, `-`, `*`, `/`, `sqrt`)
- the evaluation (`evaluate`) 

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 commented:
This is a small PR, but I had a surprising number of debates with myself with it. Some of these are more pedantic than others, but considering this is what people will be interfacing with and including in their own files, I'm looking for opinions on any of these:
- **Should the header include `Tensor.hpp`?**
- **Should the header include `TensorExpression.hpp`?** I included it so that the `TensorIndex`s are included, but these could be factored out into their own file, and then this new file could be included, instead. i.e. To write a tensor equation, you need the generic indices (`TensorIndex`) but you don't need the `TensorExpression` base class.
- **Is it appropriate to call the collective `TensorExpression` files a library?** File description may sound misleading
- **Is the name of this file what we want? (my biggest question)** Is dropping the `s` in `Expressions.hpp` preferred? I kept it plural because it's consistent with the directory and namespace `TensorExpressions`, as well as the fact that this file includes the stuff for interfacing with all concrete `TensorExpression` types. `TensorExpression.hpp` seems like the best name for this header file, but it's already in use for defining the `TensorExpression` class, and I don't think it would make sense to reserve `TensorExpression.hpp` for this header file name and then define the `TensorExpression` class in a file named differently. Not trying to fret over naming unnecessarily, but since people will be including this file, I want it to be intuitive and consistent with any naming conventions elsewhere in SpECTRE, as well as not confusing with respect to other names in the implementation, like the `TensorExpression` class , its base class, and the `TensorExpressions` namespace. I also considered naming it `TensorExpressions.hpp`, but thought it would be annoying if someone forgot the `s` and instead included the base class file `TensorExpression.hpp` and it took a while to see the mistake, but it's nice that it's consistent with the name of the base type and what we would call them: "tensor expressions."

nilsdeppe commented:
All great questions! I think the big one for me is:
Should `#include "DataStructures/Tensor/Tensor.hpp"` allow you to use the expressions fully?

If the answer is "yes" then I think the right thing to do would be one of the following:
1. have an "all includes" inside the `TensorExpressions` directory
2. just include everything in the `TensorExpressions` directory directly in `Tensor.hpp`

If the answer is "no" then I guess we should try to remove the current TensorExpression includes from `Tensor.hpp` and only have a forward declaration. This might be tricky.

I'm in favor of just doing 2. I think that's easiest on users and on maintainers :)

macedo22 commented:
@nilsdeppe For the "yes" to including `Tensor.hpp` giving you all TE functionality, I think my only question would be - could you envision a case when someone just needs `Tensor` but they would be bothered or inconvenienced that all the TE stuff was also being included and compiled? I can't think of one myself, but just wanted to make sure I asked. Otherwise, 1 and 2 both sound fine to me.

nilsdeppe commented:
I'm sure there are cases where you only need `Tensor` and nothing else, but similar to with `DataVector` it's easier for users to supply everything in the one header file than to give a bunch of different headers with "If you need X then include A,B,C, if you need Y include B,D,E" (of course not that scattered here)

macedo22 commented:
Sounds good. For option 2, for the TE files that `Tensor.hpp` doesn't actually use, is there a good way to essentially note to maintainers "hey this TE include isn't used here but don't remove it" or is that a little unnecessary? Not sure if there are other places in SpECTRE that do something like this or if there is some convention

nilsdeppe commented:
That's what I don't like about 2. :) I think it's better to include all of TEs inside `Tensor.hpp` so users get the whole package. I guess otherwise there could be documentation on `Tensor` of like "You would need to include header X.hpp to get all of the tensor expressions". But again, I think it's more user-friendly to include all the expressions in `Tensor.hpp`

macedo22 commented:
Oh sorry, I think I might not have been clear. I mean to say that 2 would mean including these in `Tensor.hpp`
```
#include "DataStructures/Tensor/Expressions/AddSubtract.hpp"  // not used by Tensor.hpp, but needs to stay
#include "DataStructures/Tensor/Expressions/Contract.hpp"
#include "DataStructures/Tensor/Expressions/Evaluate.hpp"  // not used by Tensor.hpp, but needs to stay
#include "DataStructures/Tensor/Expressions/Product.hpp"  // not used by Tensor.hpp, but needs to stay
#include "DataStructures/Tensor/Expressions/SquareRoot.hpp"  // not used by Tensor.hpp, but needs to stay
#include "DataStructures/Tensor/Expressions/TensorAsExpression.hpp"
#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
```

and I'm wondering about the case where a maintainer comes across the `Tensor.hpp` file and is like, "wait, `Tensor.hpp` doesn't need these 4 includes that it's not using" and then want to remove them. I'm wondering if there is a convention used in SpECTRE to note `include`s that aren't necessary for the file, but need to stay, or if this is a really unnecessary measure to take.

So I'm still on board with 2, just wondering if I need to add note(s) for maintainers or something

nilsdeppe commented:
oh, I see. We don't have a convention for this, but something like `// provided as TensorExpressions interface` would be fine, either on each line of an include in that category or after the includes you could add a similar comment but a bit longer. E.g.
```cpp
// Not all the TensorExpression includes are necessary, but we include them so that 
// Tensor.hpp provides a uniform interface to the Tensor and TensorExpressions
```

macedo22 commented:
@nilsdeppe Since this PR concerns what including `Tensor.hpp` means, the new first commit also updates any test files that write equations with TEs to just include `Tensor.hpp`. I took the opportunity in the second commit to do an overdue cleanup of all TE includes.

Since we've been discussing the idea of `Tensor.hpp`  being one interface to both `Tensor`s and `TensorExpression`s, I wasn't sure if for test files that create `Tensor`s with `Symmetry`, `IndexType`s, `UpLo`s, and `Frame`s, if `Symmetry.hpp` and `IndexType.hpp` should be explicitly included, or if those two files should, like the TE interface, be considered part of the Tensor interface and "come along with `Tensor.hpp`". For example, most of the TE test files create `Tensor`s like this:
```
Tensor<DataType, Symmetry<2, 1>,
         index_list<SpatialIndex<3, UpLo::Up, Frame::Inertial>,
                    SpatialIndex<3, UpLo::Lo, Frame::Inertial>>>
      Rul(used_for_size);
```

Should `Symmetry.hpp` and `IndexType.hpp` be explicitly included in test files that have this (i.e. when people are interfacing with `Tensor`s), or is including `Tensor.hpp` sufficient as we consider the two former files to be part of the Tensor interface when using `Tensor`s? If they should be explicitly included, I can add a fixup commit for this in the test files that create `Tensor`s in this way.

Sorry if this is nitpicky! Since we're on the topic of a uniform tensor interface, I thought this was relevant to address.

macedo22 commented:
Hmm looks like I mistakenly forgot to run the `Test_TensorAsExpression.cpp` test locally before pushing my second commit, and it appears I introduced a circular dependency that the fixup commit fixes. Even though the fixup fixes the compile error, I'd prefer to have `TensorAsExpression.hpp` included in files in the `TensorAsExpressionRank#TestHelpers.hpp` files, because I'm operating under the assumption that `Tensor.hpp` only includes `TensorExpression` files necessary for writing the equations. In other words, the only reason `TensorAsExpression.hpp` is currently included in `Tensor.hpp` now is because `Tensor.hpp` uses it - if it didn't, I wouldn't have it included, because it's not necessary for interfacing. I'm not sure if I should just include all the `Tensor/Expressions/...hpp` files in `Tensor.hpp` or if it's better the way I currently have it, where `Tensor.hpp` just includes what it personally uses and what is necessary to write equations. Let me know! If I do the former, I'll add the other includes to `Tensor.hpp` and remove the `TensorAsExpression.hpp` includes from the other `TensorAsExpressionRank#TestHelpers.hpp` files.

nilsdeppe commented:
Sorry for the delay, this slipped by me...

So I personally like the idea of including all the `Tensor/TensorExpressions/*.hpp` files in `Tensor.hpp` so that one only needs to include `Tensor.hpp` and gets everything. In terms of cyclic deps, I think you can get away with a forward decl in `TensorAsExpression.hpp` because that only stores a pointer and so doesn't need to know the size. Then you operate under the assumption that the user has included `Tensor.hpp` when including `TensorAsExpression.hpp`. What do you think?

macedo22 committed:
Include TensorExpression files in Tensor.hpp

In test files where TensorExpression equations are written, remove individual TensorExpression includes in lieu of just providing Tensor.hpp

macedo22 commented:
@nilsdeppe Sorry for the larger delay from me!

I think what I originally wanted was for `Tensor.hpp` to only include `Tensor/Expressions/` files that: (i) it uses itself and/or (ii) provide the interface for writing equations with TEs (i.e. anything a user would need, but not more). Here, (ii) would include the files that provide the operators, `TensorIndex`s, and `evaluate`: `AddSubtract.hpp` (`+`, `-`), `Evaluate.hpp` (`evaluate`), `Product.hpp` (`*`, `/`), `SquareRoot.hpp` (`sqrt`), and `TensorExpression.hpp` (`TensorIndex`s). The motivation for this was simply because none of the other TE files are needed for interfacing, so they don't _need_ to be included in order to write the equations with TEs. However, I'm fine just including all of the TE files wholesale if it keeps things simpler, and that's what I've done in the new first commit.

I was going to push fixups, but it started getting messy, so I restarted this work from scratch. It still remains that the first commit updates `Tensor.hpp` to include TE files and has TE test files just include `Tensor.hpp` in lieu of including individual TE files, and the second commit still cleans up the `#include`s in all the TE source and test files. I have the TE test files still include `IndexType.hpp` and `Symmetry.hpp` even when `Tensor.hpp` is included, but let me know if you think these two should just be assumed to be included as part of including `Tensor.hpp` (similar to how we will treat TE files as coming along with `Tensor.hpp`), and I can remove them.

Also, if you think the cleanup in the second commit belongs in a separate PR, let me know and I'll remove it altogether. I lumped it in since this PR mostly focuses on `#include` updates for TEs.

macedo22 committed:
Clean up includes for TensorExpressions and their tests

nilsdeppe commented on src/DataStructures/Tensor/Tensor.hpp:
```diff
@@ -15,9 +15,17 @@
+#include "DataStructures/Tensor/Expressions/SquareRoot.hpp"
 #include "DataStructures/Tensor/Expressions/TensorAsExpression.hpp"
 #include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+// Not all the TensorExpression includes are necessary, but we include them so
```
Please move this comment to after all the includes. Clang-format won't sort includes across comments, or at least older versions don't

macedo22 replied:
Done!

nilsdeppe reviewed (commented):
You can squash immediately :)
